mvrx-core/src/main/java/com/airbnb/mvrx/MavericksRepository.kt

```diff
@@ -0,0 +1,419 @@
+        if (blockExecutions != MavericksBlockExecutions.No) {
+            if (blockExecutions == MavericksBlockExecutions.WithLoading) {
+                setState { reducer(Loading(value = retainValue?.get(this)?.invoke())) }
+            }
+            // Simulate infinite loading
+            return coroutineScope.launch { delay(Long.MAX_VALUE) }
+        }
+
+        setState { reducer(Loading(value = retainValue?.get(this)?.invoke())) }
+
+        return catch { error -> setState { reducer(Fail(error, value = retainValue?.get(this)?.invoke())) } }
+            .onEach { value -> setState { reducer(Success(value)) } }
+            .launchIn(coroutineScope + (dispatcher ?: EmptyCoroutineContext))
+    }
+
+    /**
+     * Collect a Flow and update state each time it emits a value. This is functionally the same as wrapping onEach with a setState call.
+     *
+     * @param dispatcher A custom coroutine dispatcher that the coroutine will run on. If null, uses the dispatcher in [coroutineScope],
+     *                  which defaults to [Dispatchers.Main.immediate] and can be overridden globally with [Mavericks.initialize].
+     * @param reducer A reducer that is applied to the current state and should return the new state. Because the state is the receiver
+     *                and is likely a data class, an implementation may look like: `{ copy(response = it) }`.
+     */
+    protected open fun <T> Flow<T>.setOnEach(
+        dispatcher: CoroutineDispatcher? = null,
+        reducer: S.(T) -> S
+    ): Job {
+        val blockExecutions = onExecute(this@MavericksRepository)
+        if (blockExecutions != MavericksBlockExecutions.No) {
+            // Simulate infinite work
+            return coroutineScope.launch { delay(Long.MAX_VALUE) }
+        }
+
+        return onEach {
+            setState { reducer(it) }
+        }.launchIn(coroutineScope + (dispatcher ?: EmptyCoroutineContext))
+    }
+
+    /**
+     * Called each time a [execute] function is invoked. This allows
+     * the execute function to be skipped, based on the returned [MavericksBlockExecutions] value.
+     *
+     * This is intended to be used to allow the [MavericksRepository] to be mocked out for testing.
+     * Blocking calls to execute prevents long running asynchronous operations from changing the
+     * state later on when the calls complete.
+     *
+     * Mocking out the state store cannot accomplish this on its own, because in some cases we may
+     * want the state store to initially be mocked, with state changes blocked, but later on we may
+     * want it to allow state changes.
+     *
+     * This prevents the case of an executed async call from modifying state once the state stored
+     * is "enabled", even if the execute was performed when the state store was "disabled" and we
+     * didn't intend to allow operations to change the state.
+     */
+    protected open fun <S : MavericksState> onExecute(
```

Comment:
I don't follow this change. To maintain previous behavior, doesn't this need to call `config.onExecute`?

why open this up for override?

Reply:
My bad, left from couple iterations of refactoring. Addressed in https://github.com/airbnb/mavericks/pull/635/commits/3f768692c00a06d0030eece840aea79d0f5eeddb